src/Codeception/Lib/InnerBrowser.php

```diff
@@ -379,24 +376,48 @@ protected function clickByLocator($link)
         }
 
         foreach ($nodes as $node) {
-            $tag = $node->nodeName;
+            $tag = $node->tagName;
             $type = $node->getAttribute('type');
+
             if ($tag === 'a') {
-                $this->crawler = $this->clientClick($nodes->first()->link());
-                $this->forms = [];
-                break;
+                $this->amOnPage($node->getAttribute('href'));
+                return true;
             } elseif (in_array($tag, ['input', 'button']) && in_array($type, ['submit', 'image'])) {
-                $buttonValue = [];
-                if (strval($nodes->first()->attr('name')) !== '' && $nodes->first()->attr('value') !== null) {
-                    $buttonValue = [$nodes->first()->attr('name') => $nodes->first()->attr('value')];
-                }
+                return $this->clickButton($node);
+            }
+        }
+    }
+
+
+    /**
+     * Clicks the link or submits the form when the button is clicked
+     * @param \DOMNode $node
+     * @return boolean clicked something
+     */
+    private function clickButton(\DOMNode $node)
+    {
+        $formParams = [];
+        $buttonName = (string)$node->getAttribute('name');
+        $buttonValue = $node->getAttribute('value');
+
+        if ($buttonName !== '' && $buttonValue !== null) {
+            $formParams = [$buttonName => $buttonValue];
+        }
+
+        while ($node->parentNode !== null) {
+            $node = $node->parentNode;
+            if ($node->tagName === 'a') {
+                $this->amOnPage($node->getAttribute('href'));
```

Comment:
Just wondering should this behavior be applied to any element? If it is a part of a link clicking on it should move to corresponding page. How do you think?

Also what if href=='#'? This will produce a failure.